### PR TITLE
Jsonnet fmt execution at save

### PR DESCRIPTION
**Pull request by ahakanbaba**
This change adds auto jsonnet fmt execution at save time for *.libsonnet and *.jsonnet files.  This fixes the issue #2 

(The inspiration comes from auto gofmt execution in vim-go plugin. )

There is a vim command and also a user visible function that can be used to execute `jsonnet fmt`. All documented in the appropriate document. 

We have used reasonable defaults for the jsonnet fmt and added configurability. 

Our local manual tests used only osX environment. This "should" work in a linux box as is. However it would definitely not work in windows. We have not attempted to handle windows platform because we do not have a machine to test on. If this gets accepted, we will create an issue about this. 

This change unfortunately lacks automated tests. If anyone has some experience or suggestions how to write tests for vim scripts we would appreciate very much.

### CODE REVIEW

**Comment by ahakanbaba** [comment]
The auto execution at save is enabled by default. 
This is also same as in vim-go. 
Once one sources vim-go, the auto execution of GoFmt at save time is enabled by default. 
This change has the same behavior. 

For manual tests, one might prefer not to pollute the original ~/.vim directory. 
I can suggest to download the repo in a separate directory and then launching the vim with the following option 

vim --cmd 'set rtp^=/Users/hbaba/src/vim-jsonnet'

This includes the /Users/hbaba/src/vim-jsonnet to the RunTimePath of vim. Assuming the /Users/hbaba/src/vim-jsonnet is the directory where you have downloaded the latest code.

**Comment by ahakanbaba** [comment]
@sparkprime  Hi David, I would appreciate if you could take a look at this "Jsonnet fmt execution at save PR" at some point when you are available. 
Thanks
Hakan

**Comment by ahakanbaba** [comment]
Hi @sparkprime, do you think you could take a look at this anytime soon ? 
We would love to get this merged into the master branch so that we can more easily benefit from it.

**Comment by sparkprime** [comment]
Sorry I didn't get back.  Yes, I will look at it next week.  I've been on vacation the last couple of weeks so only replying to github and not actually doing anything :)  thanks

**Comment by ahakanbaba** [comment]
Hi David, 
thanks for the heads-up. 
Looking forward to your comments.

**Comment by ahakanbaba** [comment]
@sparkprime 
Hi, 
I wanted to ping you again about this pull request. 
I would really appreciate if you can take a look at this. 
Thanks

File: autoload/jsonnet.vim [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    let l:output = call("system", [a:str] + a:000)
+    return l:output
+  finally
+    let &shell = l:shell
+  endtry
+endfunction
+
+
+" CheckBinPath checks whether the given binary exists or not and returns the
+" path of the binary. It returns an empty string doesn't exists.
```
Comment (sparkprime) [inline autoload/jsonnet.vim]: s/doesn't exists/if it doesn't exist/

File: autoload/jsonnet.vim [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+   " The inplace modification is default. Makes file management easier
+   " Indentation spacing is 4 by default
+   let l:command = l:command . ' -i -n 4 '
+   let l:command = l:command . g:jsonnet_fmt_options
+
+   " Execute the compiled jsonnet fmt command and save the return value
+   let l:out = jsonnet#System(l:command . " " . l:tmpname)
+   let l:errorCode = v:shell_error
+
+   " The format command succeeded
```
Comment (sparkprime) [inline autoload/jsonnet.vim]: Please move comment inside the true branch

File: doc/vim-jsonnet.txt [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+                                                         *'g:jsonnet_fmt_on_save'*
+
+Use this option to auto |:JsonnetFmt| on save. By default it's enabled >
+
+  let g:jsonnet_fmt_on_save = 1
+<
+                                                             *'g:jsonnet_command'*
+
+Use this option to define which tool is used to jsonet fmt. By default `jsonnet` is
```
Comment (sparkprime) [inline doc/vim-jsonnet.txt]: s/jsonet fmt/format/

File: doc/vim-jsonnet.txt [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+used >
+
+  let g:jsonnet_command = "jsonnet"
+<
+                                                         *'g:jsonnet_fmt_command'*
+
+Use this option to define which <cmd> parameter is used to *g:jsonnet_command* tool.
```
Comment (sparkprime) [inline doc/vim-jsonnet.txt]: s/used to/used with/

File: doc/vim-jsonnet.txt [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+By default `fmt` isused >
```
Comment (sparkprime) [inline doc/vim-jsonnet.txt]: s/isused/is used/

File: doc/vim-jsonnet.txt [hunk 6]
```diff
@@ -0,0 +1,105 @@
+*vim-jsonnet.txt* Jsonnet development plugin
+*vim-jsonnet*
+
+
```
Comment (sparkprime) [inline doc/vim-jsonnet.txt]: Lots of whitespace at the top, and I guess you want some under the banner?  I presume this is not a requirement of the format of this file (I've never written vim docs before).

File: doc/vim-jsonnet.txt (hunk 5 above)
Comment (ahakanbaba) [inline doc/vim-jsonnet.txt]: Thanks, I have addressed these comments. 
Special thanks for the diligent review paying attention to typos.

**Comment by ahakanbaba** [comment]
Hi @sparkprime 
I have addressed the typos and other issues you have pointed out. 
Please feel free to take another look.

**Comment by ahakanbaba** [comment]
@googlebot 
I am the only author in these commits.

**Comment by benley** [comment]
@ahakanbaba googlebot thinks there are multiple authors because your branch includes a cherry-picked copy of that syntastic integration change.  If you rebase your branch against the current master and force-push it, that issue should disappear.

**Comment by ahakanbaba** [comment]
@sparkprime @benley.
Is there anything else I need to do? Is there any other action items on me that block the merge ? 

I have done the force push and rebase mentioned [in a previous comment](https://github.com/google/vim-jsonnet/pull/3#issuecomment-252064247) although I do not fully understand the confusion of the googleBot. 

I have not done any manual cherrypicking on my branch.

In any case does everything look good to you ?

**Comment by sparkprime** [comment]
You still have a commit from @benley in this PR (check "commits" at the top), which for posterity is a commit from another PR and therefore we already have a record of consent for it.  But it's standard practice to force push a rebased change where everything is squashed into a single commit, just so the commit log on HEAD is less fractured.  Doing that would remove the benley commit (and the related changes because they're already in HEAD).

git checkout master
git pull upstream master
git checkout jsonnetFmtAtSave
git rebase -i master
squash everything into one commit
rewrite commit message to be a short summary
git push -f

When you've done that, the "commits" tab at the top will show only 1 commit and this will also kick GoogleBot to OK the PR.

**Comment by ahakanbaba** [comment]
ok Thanks Dave,
I will do that today.
Hakan

**Pull request closed without merging**